src/assign.c

```diff
@@ -911,16 +871,41 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
         memcpy(RAW(target)+start, RAW(source), slen*SIZEOF(target));
       }
       break;
-    case LGLSXP: case INTSXP :
-      if (TYPEOF(source)!=LGLSXP && TYPEOF(source)!=INTSXP) { source = PROTECT(coerceVector(source, TYPEOF(target))); protecti++; }
+    case LGLSXP:
+    case INTSXP: {
+      int *td = INTEGER(target)+start;
       if (slen==1) {
-        int *td = INTEGER(target)+start;
-        const int val = INTEGER(source)[0];
-        for (int i=0; i<len; i++) td[i] = val;
+        // punned-coerce; i.e. avoid the small overhead of allocating via coerceVector
+        int val;
+        switch (TYPEOF(source)) {
+        case LGLSXP:
+        case INTSXP:
+          val = INTEGER(source)[0];
+          break;
+        case REALSXP: {   // allowed real 0/1/NA for logical, and no float for integer, was checked up front above
+          double d = REAL(source)[0];
+          val = ISNAN(d) ? NA_INTEGER : (int)d;
+        } break;
+        default:
+          error("Internal error"); // # nocov
+        }
+        for (int i=0; i<len; ++i) td[i] = val;  // recycle length-1
       } else {
-        memcpy(INTEGER(target)+start, INTEGER(source), slen*SIZEOF(target));
+        switch (TYPEOF(source)) {
+        case LGLSXP:
+        case INTSXP:
+          memcpy(td, INTEGER(source), slen*SIZEOF(target));  // correct type ideal length>1 case
+          break;
+        case REALSXP: {
+          const double *sourceD = REAL(source);
+          for (int i=0; i<len; ++i) {
```

Comment:
the `memcpy` branch is more efficient, is that right? should we report something under `verbose` then?

Reply:
then we can also test that branch is reached exactly when we expect that by testing verbose output.

Reply:
yep there's a todo comment in the place to add the verbose/strict option.  Can do it in the next release.